### PR TITLE
look at the body to determine if it has content or not

### LINKED ISSUE #148: Transfer-Encoding: chunked does not imply a body is present
Consider a service written with `go-openapi` receiving an HTTP request with the following attributes:
- GET method
- No content-type header
- No content-length header
- A `transfer-encoding: chunked` header
- A single, empty chunk (ie: the body transmitted is semantically empty, but does consume bytes on the HTTP body as is consistent with chunked encoding)

I see in practice that the code assumes the default `content-type: application/octet-stream` for this request because it believes there is a body. By assuming the default content type, service specs that do not explicitly support octet-stream will reject this request as invalid (HTTP 415). I believe the code responsible for deciding whether a request gets a content-type based on the presence of a request body is here: https://github.com/go-openapi/runtime/blob/master/request.go#L45

I've been struggling to definitively prove or disprove whether this behavior is strictly correct according to https://tools.ietf.org/html/rfc2616#section-14.17 and https://tools.ietf.org/html/rfc2616#section-14.41

Specifically, I interpret the RFC as saying:
- Transfer encoding and content length are mutually exclusivel. You can choose to use one or the other without changing the semantics of the message.
- Transfer encoding is a property of the message, not the entity.
- Content-Type is a property of the entity.

Reviewing https://github.com/go-openapi/runtime/blob/master/request.go#L45 it feels like..

// HasBody returns true if this method needs a content-type
func HasBody(r *http.Request) bool {
	return len(r.TransferEncoding) > 0 || r.ContentLength > 0
}

..is suggesting that any request with a non-zero content length has a body. I believe this is correct. However, it also suggests that any request with a transfer-encoding header has a body and, according to the comment, "needs a content-type". Since the entity can still be empty (ie: a single empty chunk), and empty entities do not require a content-type, this seems wrong.

I'd like to start a discussion on whether this is a bug, or if there is additional nuance in the RFC or other practicalities that make the current behavior desirable.

### PR DESCRIPTION
**Pull request by casualjim**
fixes #148

Signed-off-by: Ivan Porto Carrero <ivan@flanders.co.nz>

### CODE REVIEW

File: request.go [hunk 1]
```diff
@@ -42,7 +43,68 @@ func AllowsBody(r *http.Request) bool {
...
+}
+
+func (p *peekingReader) HasContent() bool {
+	if p == nil {
+		return false
+	}
+	if p.underlying.Buffered() > 0 {
+		return true
+	}
+	b, err := p.underlying.Peek(1)
```
Comment (esmet) [inline request.go]: Will this return pre-decoded bytes? If so, we may need to be careful about the the fact that an empty entity-body in `chunked` transfer-encoding is represented as a single ascii "0" after headers, and before the terminating CRLF.

Something like:
```
..headers...
transfer-encoding: chunked\r\n
\r\n
0\r\n
```

If this `underlying` was given bytes that already handles the transfer "decoding" then we should be safe here, but I can't tell exactly from the surrounding code just yet.

File: request.go [hunk 2]
```diff
@@ -42,7 +43,68 @@ func AllowsBody(r *http.Request) bool {
...
+		return true
+	}
+
+	if r.Header.Get(http.CanonicalHeaderKey("content-length")) != "" {
+		// in this case, no Transfer-Encoding should be present
+		// we have a header set but it was explicitly set to 0, so we assume no body
+		return false
+	}
+
+	rdr := newPeekingReader(r.Body)
```
Comment (esmet) [inline request.go]: my concern below is invalid if `r.Body` represents an entity-body after all transfer-encodings have been applied. but if it does not, then I think the `peek(1)` will observe ascii "0" (end of chunk marker) and mistakenly assume there is a body.

File: request.go (hunk 1 above)
Comment (esmet) [reply request.go]: The Go documentation is slightly comforting, but not specific:

```
 178         // For server requests, the Request Body is always non-nil
 179         // but will return EOF immediately when no body is present.
 180         // The Server will close the request body. The ServeHTTP
 181         // Handler does not need to.
 182         Body io.ReadCloser
```

Comment (esmet) [reply request.go]: Great: https://github.com/golang/go/issues/11675#issuecomment-123850603 sounds like it handles transfer encodings, or at least it did 4 years ago.

File: request.go (hunk 2 above)
Comment (casualjim) [reply request.go]: the go stdlib takes care of this part

**Comment by jhewes** [comment]
I have ran a test with the patch and I can confirm it has resolved the issue we were seeing.

**Pull request merged**